### ISSUE TITLE
🚀 Context-aware line wrapping

### ISSUE THREAD

**Issue by zachriggle** [dandavison/delta#657]
### Preface

Once again, `delta` is an AMAZING tool that I use every day, and definitely makes my workflows (lots of diff reviews) much simpler (and better than GitHub / BitBucket / GitLab diff viewers, due to supporting all of Git's options e.g. `-W`).

### Problem Description

One of the issues that arises frequently is that some of our Java and Swift programmers like having lines that are 600 characters long.

Since `delta` uses one of many various pagers, usually some form of `less`, which supports line-wrapping, but this breaks lots of things (like `delta`'s line numbers).  Specifying `LESS=... -S ...` allows side-scroll to view the rest of the line, but this is quite tedious.

This becomes a double-whammy when using `delta --side-by-side` as the long lines are simply truncated, especially on the left-hand side (before).  Scrolling only reveals more on the right-hand (after) side.

### Feature Request

Ultimately, it would be nice if `delta` had its own option to support wrapping lines in some kind of context-aware manner.  "Context-aware" can simply be "wrap lines, and prepend N+4 spaces" where "N" is the number of leading spaces in the original line.  There would likely need to be actual "context aware magic" so that the full wrapped line appears to be on the same line number with `delta -n`, but I assume this is achievable.

This would also make `--side-by-side` more useful, if applied to both sides, especially with long lines.

### Pitfalls and Non-Goals

This will still fail in predictable ways when the number of leading whitespace is large, as apparently some people prefer:

```c
if(...) { 
  if(...) {
    if(...) {
      if(...) {
        if(...) {
           // Ad infinitum, until eventually....
           main_logic();
        } else {
            return err;
        }
      } else {
          return err;
      }
    } else {
        return err;
    }
  } else {
    return err2;
  }
} else {
  return err1;
}
```

Instead of the (IMHO more readable)

```c
if(!...) { 
  return err1;
}

if(!...) { 
  return err2;
}

// add infinitum

main_logic();
```

**Comment by dandavison** [comment]
Hi @zachriggle, for side-by-side there is good news: @th1000s has a branch in an advanced state that implements quite sophisticated line wrapping in side-by-side mode: https://github.com/dandavison/delta/pull/515 

I wonder whether it might be possible, after that merges, to borrow the line-wrapping technology from that work and use it for the non-side-by-side case; I don't think that's been discussed yet.

**Comment by th1000s** [comment]
That sounds doable. Before finishing #515 maybe the `--side-by-side-wrap-symbol` and related arguments should be renamed to `--wrap-symbol`. Two specific flags, the original and `--unified-wrap-symbol`, could always be added later.

**Comment by dandavison** [comment]
@th1000s I agree: removing the `side-by-side` prefix from some option names seems to have several advantages (e.g. brevity, generality) and as you say we can always introduce the s-b-s specific ones later. Are you thinking of removing the prefix from all the options below? 

```
        --side-by-side-wrap-max-lines <side-by-side-wrap-max-lines>
            Maximum number of wrapped lines to display in side-by-side mode. Any content that still does not fit will be
            truncated [default: 3]
        --side-by-side-wrap-symbol <side-by-side-wrap-symbol>
            Symbol indicating that a line has been wrapped in side-by-side mode [default: ↵]
        --side-by-side-wrap-right-percent <side-by-side-wrap-right-percent>
            Threshold for right-aligning wrapped content in side-by-side mode. If the length of remaining wrapped
            content, as a percentage of the panel width, is less than this quantity then it will be displayed right-
            aligned [default: 37.0]
        --side-by-side-wrap-right-wrap-symbol <side-by-side-wrap-right-wrap-symbol>
            Symbol indicating that a line has been wrapped and that the subsequent content is displayed right-aligned
            [default: ↴]
        --side-by-side-wrap-right-prefix-symbol <side-by-side-wrap-right-prefix-symbol>
            Symbol displayed in front of right-aligned wrapped content [default: …]
```

**Comment by th1000s** [comment]
Yes, and I removed the prefix in the other PR, which is now final.
